// @owlbear-rodeo/sdk/lib/api/AssetsApi.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
class AssetsApi {
    constructor(messageBus) {
        this.messageBus = messageBus;
    }
    uploadImages(images, typeHint) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_ASSETS_UPLOAD_IMAGES", {
                images,
                typeHint,
            });
        });
    }
    uploadScenes(scenes, disableShowScenes) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_ASSETS_UPLOAD_SCENES", {
                scenes,
                disableShowScenes,
            });
        });
    }
    downloadImages(multiple, defaultSearch, typeHint) {
        return __awaiter(this, void 0, void 0, function* () {
            const { images } = yield this.messageBus.sendAsync("OBR_ASSETS_DOWNLOAD_IMAGES", { multiple, defaultSearch, typeHint }, -1);
            return images;
        });
    }
    downloadScenes(multiple, defaultSearch) {
        return __awaiter(this, void 0, void 0, function* () {
            const { scenes } = yield this.messageBus.sendAsync("OBR_ASSETS_DOWNLOAD_SCENES", { multiple, defaultSearch }, -1);
            return scenes;
        });
    }
}
export default AssetsApi;
