// @owlbear-rodeo/sdk/lib/api/PopoverApi.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { normalizeUrlObject } from "../common/normalize";
class PopoverApi {
    constructor(messageBus) {
        this.messageBus = messageBus;
    }
    open(popover) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_POPOVER_OPEN", Object.assign({}, normalizeUrlObject(popover)));
        });
    }
    close(id) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_POPOVER_CLOSE", { id });
        });
    }
    getWidth(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const { width } = yield this.messageBus.sendAsync("OBR_POPOVER_GET_WIDTH", { id });
            return width;
        });
    }
    setWidth(id, width) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_POPOVER_SET_WIDTH", { id, width });
        });
    }
    getHeight(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const { height } = yield this.messageBus.sendAsync("OBR_POPOVER_GET_HEIGHT", { id });
            return height;
        });
    }
    setHeight(id, height) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.messageBus.sendAsync("OBR_POPOVER_SET_HEIGHT", { id, height });
        });
    }
}
export default PopoverApi;
