// @owlbear-rodeo/sdk/lib/builders/SceneUploadBuilder.js
export class SceneUploadBuilder {
    constructor() {
        this._upload = {
            name: "New Scene",
            fog: { filled: false, style: { color: "#222222", strokeWidth: 5 } },
            grid: {
                dpi: 150,
                scale: "5ft",
                style: { lineColor: "LIGHT", lineOpacity: 0.4, lineType: "DASHED" },
                measurement: "CHEBYSHEV",
                type: "SQUARE",
            },
            items: [],
        };
    }
    name(name) {
        this._upload.name = name;
        return this;
    }
    fogFilled(filled) {
        this._upload.fog.filled = filled;
        return this;
    }
    fogColor(color) {
        this._upload.fog.style.color = color;
        return this;
    }
    fogStrokeWidth(width) {
        this._upload.fog.style.strokeWidth = width;
        return this;
    }
    gridScale(scale) {
        this._upload.grid.scale = scale;
        return this;
    }
    gridColor(color) {
        this._upload.grid.style.lineColor = color;
        return this;
    }
    gridOpacity(opacity) {
        this._upload.grid.style.lineOpacity = opacity;
        return this;
    }
    gridLineType(lineType) {
        this._upload.grid.style.lineType = lineType;
        return this;
    }
    gridMeasurement(measurement) {
        this._upload.grid.measurement = measurement;
        return this;
    }
    gridType(type) {
        this._upload.grid.type = type;
        return this;
    }
    items(items) {
        this._upload.items = items;
        return this;
    }
    baseMap(baseMap) {
        this._upload.baseMap = baseMap;
        return this;
    }
    thumbnail(thumbnail) {
        this._upload.thumbnail = thumbnail;
        return this;
    }
    build() {
        return this._upload;
    }
}
